File: src/main/python/AoC2023_17.py
#! /usr/bin/env python3
#
# Advent of Code 2023 Day 17
#

import sys
from collections import defaultdict
from queue import PriorityQueue

from aoc.common import InputData
from aoc.common import SolutionBase
from aoc.common import aoc_samples
from aoc.grid import IntGrid

Input = IntGrid
Output1 = int
Output2 = int


TEST = """\
2413432311323
3215453535623
3255245654254
3446585845452
4546657867536
1438598798454
4457876987766
3637877979653
4654967986887
4564679986453
1224686865563
2546548887735
4322674655533
"""

DIRS = {(0, 1), (0, -1), (1, 0), (-1, 0)}
Move = tuple[int, int, int, int]


class Solution(SolutionBase[Input, Output1, Output2]):
    def parse_input(self, input_data: InputData) -> Input:
        return IntGrid.from_strings([line for line in input_data])

    def solve(self, grid: Input, min_moves: int, max_moves: int) -> int:
        max_r = grid.get_height() - 1
        max_c = grid.get_width() - 1

        q: PriorityQueue[tuple[int, Move]] = PriorityQueue()
        q.put((0, (0, 0, 0, 0)))
        best: defaultdict[Move, int] = defaultdict(lambda: int(1e9))
        best[(0, 0, 0, 0)] = 0
        while not q.empty():
            cost, move = q.get()
            r, c, dr, dc = move
            if r == max_r and c == max_c:
                return cost
            heatloss = best[move]
            for drr, dcc in DIRS - {(dr, dc), (-dr, -dc)}:
                rr, cc, hl = r, c, 0
                for i in range(1, max_moves + 1):
                    rr += drr
                    cc += dcc
                    if not (0 <= rr <= max_r and 0 <= cc <= max_c):
                        break
                    hl += grid.values[rr][cc]
                    if i >= min_moves:
                        n = (rr, cc, drr, dcc)
                        new_heatloss = heatloss + hl
                        if new_heatloss < best[n]:
                            best[n] = new_heatloss
                            q.put((new_heatloss, n))
        raise RuntimeError("unsolvable")

    def part_1(self, grid: Input) -> Output1:
        return self.solve(grid, 1, 3)

    def part_2(self, grid: Input) -> Output2:
        return self.solve(grid, 4, 10)

    @aoc_samples(
        (
            ("part_1", TEST, 102),
            ("part_2", TEST, 94),
        )
    )
    def samples(self) -> None:
        pass


solution = Solution(2023, 17)


def main() -> None:
    solution.run(sys.argv)


if __name__ == "__main__":
    main()

File: src/main/java/AoC2023_17.java
import java.util.List;
import java.util.Set;
import java.util.function.Function;
import java.util.stream.Stream;
import java.util.stream.Stream.Builder;

import com.github.pareronia.aoc.Grid.Cell;
import com.github.pareronia.aoc.IntGrid;
import com.github.pareronia.aoc.SetUtils;
import com.github.pareronia.aoc.geometry.Direction;
import com.github.pareronia.aoc.geometry.Turn;
import com.github.pareronia.aoc.graph.Dijkstra;
import com.github.pareronia.aoc.solution.Sample;
import com.github.pareronia.aoc.solution.Samples;
import com.github.pareronia.aoc.solution.SolutionBase;

public final class AoC2023_17
        extends SolutionBase<IntGrid, Integer, Integer> {
    
    private AoC2023_17(final boolean debug) {
        super(debug);
    }
    
    public static AoC2023_17 create() {
        return new AoC2023_17(false);
    }
    
    public static AoC2023_17 createDebug() {
        return new AoC2023_17(true);
    }
    
    @Override
    protected IntGrid parseInput(final List<String> inputs) {
        return IntGrid.from(inputs);
    }
    
    private int solve(final IntGrid grid, final int minMoves, final int maxMoves) {
        record Move(Cell cell, Direction dir, int cost) {}
        
        final Function<Move, Stream<Move>> adjacent = move -> {
            final Set<Direction> dirs = move.dir() == null
                ? Direction.CAPITAL
                : SetUtils.difference(Direction.CAPITAL,
                    Set.of(move.dir(), move.dir().turn(Turn.AROUND)));
            final Builder<Move> moves = Stream.builder();
            for (final Direction dir : dirs) {
                Cell cell = move.cell();
                int hl = 0;
                for (int i = 1; i <= maxMoves; i++) {
                    cell = cell.at(dir);
                    if (!grid.isInBounds(cell)) {
                        break;
                    }
                    hl += grid.getValue(cell);
                    if (i >= minMoves) {
                        moves.add(new Move(cell, dir, hl));
                    }
                }
            }
            return moves.build();
        };
        final Cell end = Cell.at(grid.getMaxRowIndex(), grid.getMaxColIndex());
        return (int) Dijkstra.distance(
                new Move(Cell.at(0, 0), null, 0),
                move -> move.cell().equals(end),
                adjacent,
                (curr, next) -> next.cost);
    }

    @Override
    public Integer solvePart1(final IntGrid grid) {
        return solve(grid, 1, 3);
    }
    
    @Override
    public Integer solvePart2(final IntGrid grid) {
        return solve(grid, 4, 10);
    }
    
    @Override
    @Samples({
        @Sample(method = "part1", input = TEST1, expected = "102"),
        @Sample(method = "part2", input = TEST1, expected = "94"),
        @Sample(method = "part2", input = TEST2, expected = "71"),
    })
    public void samples() {
    }
    
    public static void main(final String[] args) throws Exception {
        AoC2023_17.create().run();
    }

    private static final String TEST1 = """
            2413432311323
            3215453535623
            3255245654254
            3446585845452
            4546657867536
            1438598798454
            4457876987766
            3637877979653
            4654967986887
            4564679986453
            1224686865563
            2546548887735
            4322674655533
            """;
    private static final String TEST2 = """
            111111111111
            999999999991
            999999999991
            999999999991
            999999999991
            """;
}
